test/types/client_bulk_write.test-d.ts

```diff
@@ -0,0 +1,258 @@
+  type UUID,
+  type WithoutId
+} from '../mongodb';
+
+declare const client: MongoClient;
+type Book = { title: string; released: Date };
+type Author = { name: string; published: number };
+type Store = { _id: UUID };
+
+// Baseline check that schema modifies the following fields for each type.
+expectType<ClientInsertOneModel<Book>['document']>(null as unknown as OptionalId<Book>);
+
+expectType<ClientReplaceOneModel<Book>['filter']>(null as unknown as Filter<Book>);
+expectType<ClientReplaceOneModel<Book>['replacement']>(null as unknown as WithoutId<Book>);
+
+expectType<ClientUpdateOneModel<Book>['filter']>(null as unknown as Filter<Book>);
+expectType<ClientUpdateOneModel<Book>['update']>(
+  null as unknown as UpdateFilter<Book> | Document[]
+);
+
+expectType<ClientUpdateManyModel<Book>['filter']>(null as unknown as Filter<Book>);
+expectType<ClientUpdateManyModel<Book>['update']>(
+  null as unknown as UpdateFilter<Book> | Document[]
+);
+
+expectType<ClientDeleteOneModel<Book>['filter']>(null as unknown as Filter<Book>);
+expectType<ClientDeleteManyModel<Book>['filter']>(null as unknown as Filter<Book>);
+
+client.bulkWrite([]); // empty should always work
+
+// No schemas - all correct use
+client.bulkWrite([
+  {
+    namespace: 'db.authors',
+    name: 'insertOne',
+    document: { name: 'bob', published: 2 }
+  },
+  {
+    namespace: 'db.authors',
+    name: 'replaceOne',
+    filter: { name: 'bob' },
+    replacement: { name: 'ann', published: 2 }
+  },
+  {
+    namespace: 'db.authors',
+    name: 'updateOne',
+    filter: { name: 'bob', published: 2 },
+    update: {}
+  },
+  {
+    namespace: 'db.authors',
+    name: 'updateMany',
+    filter: { name: 'bob', published: 2 },
+    update: {}
+  },
+  { namespace: 'db.authors', name: 'deleteOne', filter: {} },
+  { namespace: 'db.authors', name: 'deleteMany', filter: {} }
+]);
+
+// No schemas - incorrect use - random namespaces, no type checking
+client.bulkWrite([
+  {
+    namespace: 'db.whatever',
+    name: 'insertOne',
+    document: { name: 'bob', randomKey: 2 }
+  },
+  {
+    namespace: 'db.change',
+    name: 'replaceOne',
+    filter: { name: 'bob' },
+    replacement: { name: 2, published: 2 }
+  },
+  {
+    namespace: 'db.it',
+    name: 'updateOne',
+    filter: { name: 'bob', published: new Date() },
+    update: {}
+  },
+  {
+    namespace: 'db.up',
+    name: 'updateMany',
+    filter: { name: 'bob', published: 2 },
+    update: {}
+  },
+  { namespace: 'db.random', name: 'deleteOne', filter: {} },
+  { namespace: 'db.namespace', name: 'deleteMany', filter: {} }
+]);
+
+// Operation names are still type checked when there is no schema
+expectError<ClientBulkWriteModel>({
+  namespace: 'db.author',
+  name: 'insertLots', // Not an operation we support
+  document: { name: 'bob', published: 2 }
+});
+
+type MongoDBSchemas = {
+  'db.books': Book;
+  'db.authors': Author;
+  'db.stores': Store;
+};
+
+expectError<ClientBulkWriteModel<MongoDBSchemas>>({
```

Comment:
I thought we wanted this to compile but to fallback to no type safety.

Reply:
If the keys are strings, can we just fallback to Document?  

If I'm understanding correctly, the current API requires users to either completely strictly type their `bulkWrite` api (map all namespaces to all schemas) or default to document - there's no in between.

Reply:
What is the in-between use case you're thinking of?

If you add `[key: string]: Document` to your SchemaMap you remove type assertion but also typo checking, so you revert to the default experience

Can you write out the TS you're imagining?

Reply:
I'd expect users to be able to do:

```
const schema = { 
  'foo.bar': { name: string }
}

// fine
client.bulkWrite<schema>([
  { name: 'insertOne', namespace: 'foo.bar', document: { name: 'bailey' } }
])

// also fine
client.bulkWrite<schema>([
  { name: 'insertOne', namespace: 'foo.other-ns', document: { name: 'bailey' } }
])

// also fine
client.bulkWrite<schema>([
  { name: 'insertOne', namespace: 'foo.bar', document: { name: 'bailey' } },
  { name: 'insertOne', namespace: 'foo.other-ns', document: { name: 'bailey' } }
])

// error - type checking on the supplied namespaces
client.bulkWrite<schema>([
  { name: 'insertOne', namespace: 'foo.bar', document: { name: 3 } }, // error
  { name: 'insertOne', namespace: 'foo.other-ns', document: { name: 'bailey' } }
])
```

Reply:
Thanks for the example, but I was wondering what the implementation for that would look like? That behavior would remove checking for namespace typos right?

Reply:
From standup: checking the namespaces for typos is incompatible with making plain strings map to Document, so I don't think we want to do this. 

I did look into this though, fell short of solution, I was toying around with deploying something like this:
```ts
type IsStringLiteral<T> = T extends string
  ? string extends T
    ? false
    : true
  : false;
```

But I think Namespace is always a string literal since it comes from keyof